<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Copy number variation analysis with Bioconductor  • CNVWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Copy number variation analysis with Bioconductor ">
<meta property="og:description" content="This workshop gives an overview of Bioconductor solutions for the
    analysis of copy number variation (CNV) data. The workshop introduces Bioconductor 
    core data structures for efficient representation, access, and manipulation of
    CNV data, and how to use these containers for structured downstream analysis of
    CNVs and integration with gene expression and quantitative phenotypes. 
    Participants will be provided with code and hands-on practice for a comprehensive
    set of typical analysis steps including exploratory analysis, summarizing 
    individual CNV calls across a population, overlap analysis with functional 
    genomic regions and regulatory elements, expression quantitative trait loci 
    (eQTL) analysis, and genome-wide association analysis (GWAS) with quantitative 
    phenotypes. As an advanced application example, the workshop also introduces
    allele-specific absolute copy number analysis and how it is incorporated in 
    cancer genomic analysis for the estimation of tumor characteristics such as 
    tumor purity and ploidy.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">CNVWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/Geistlinger_CNVWorkshop.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/waldronlab/CNVWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="https://github.com/waldronlab/CNVWorkshop/workflows/.github/workflows/basic_checks.yaml/badge.svg"></p>
<div id="copy-number-variation-analysis-with-bioconductor" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#copy-number-variation-analysis-with-bioconductor" class="anchor"></a>Copy number variation analysis with Bioconductor</h1></div>
<p><a href="https://waldronlab.github.io/CNVWorkshop">Workshop website</a></p>
<p><a href="https://hub.docker.com/repository/docker/ludwigg/cnvworkshop">Docker image</a></p>
</div>
<div id="instructors-names-and-contact-information" class="section level1">
<h1 class="hasAnchor">
<a href="#instructors-names-and-contact-information" class="anchor"></a>Instructor(s) name(s) and contact information</h1>
<p>Ludwig Geistlinger, Marcel Ramos, Sehyun Oh, and Levi Waldron</p>
<p>CUNY School of Public Health 55 W 125th St, New York, NY 10027</p>
<p><a href="mailto:Ludwig_Geistlinger@hms.harvard.edu" class="email">Ludwig_Geistlinger@hms.harvard.edu</a> <a href="mailto:Marcel.Ramos@sph.cuny.edu" class="email">Marcel.Ramos@sph.cuny.edu</a><br><a href="mailto:Sehyun.Oh@sph.cuny.edu" class="email">Sehyun.Oh@sph.cuny.edu</a><br><a href="mailto:Levi.Waldron@sph.cuny.edu" class="email">Levi.Waldron@sph.cuny.edu</a></p>
</div>
<div id="workshop-description" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h1>
<p>This workshop gives an overview of Bioconductor solutions for the analysis of copy number variation (CNV) data. The workshop introduces Bioconductor core data structures for efficient representation, access, and manipulation of CNV data, and how to use these containers for structured downstream analysis of CNVs and integration with gene expression and quantitative phenotypes. Participants will be provided with code and hands-on practice for a comprehensive set of typical analysis steps including exploratory analysis, summarizing individual CNV calls across a population, overlap analysis with functional genomic regions and regulatory elements, expression quantitative trait loci (eQTL) analysis, and genome-wide association analysis (GWAS) with quantitative phenotypes. As an advanced application example, the workshop also introduces allele-specific absolute copy number analysis and how it is incorporated in cancer genomic analysis for the estimation of tumor characteristics such as tumor purity and ploidy.</p>
</div>
<div id="pre-requisites" class="section level1">
<h1 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h1>
<ul>
<li><p>Basic knowledge of R syntax</p></li>
<li><p>Familiarity with the SummarizedExperiment class</p></li>
<li><p>Familiarity with the GenomicRanges class</p></li>
<li><p>Familiarity with high-throughput genomic assays such as microarrays and next-generation sequencing</p></li>
<li><p>Familiarity with the biological definition of single nucleotide polymorphism (SNP) and copy number variation (CNV)</p></li>
</ul>
</div>
<div id="workshop-participation" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h1>
<p>Execution of example code and hands-on practice</p>
</div>
<div id="r--bioconductor-packages-used" class="section level1">
<h1 class="hasAnchor">
<a href="#r--bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h1>
<ul>
<li><a href="http://bioconductor.org/packages/RaggedExperiment">RaggedExperiment</a></li>
<li><a href="http://bioconductor.org/packages/CNVRanger">CNVRanger</a></li>
<li><a href="http://bioconductor.org/packages/regioneR">regioneR</a></li>
<li><a href="http://bioconductor.org/packages/PureCN">PureCN</a></li>
</ul>
</div>
<div id="time-outline" class="section level1">
<h1 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h1>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Data representation and manipulation</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Integrative downstream analysis (eQTL, GWAS, …)</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Allele-specific CN analysis in cancer</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-goals" class="section level1">
<h1 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning Goals</h1>
<ul>
<li>Gain familiarity with elementary concepts of CNV analysis</li>
<li>Learn how to efficiently represent, access, and manipulate CNV data in Bioconductor data structures</li>
<li>Gain familiarity with different strategies for summarizing individual CNV calls across a population</li>
<li>Learn how to assess the significance of overlaps between CNVs and functional genomic regions</li>
<li>Learn how carry out association analysis with gene expression and quantitative phenotypes</li>
<li>Gain familiarity with allele-specific absolute CN analysis of cancer genomic data</li>
<li>Understand how CNVs can be experimentally detected and computationally inferred from SNP arrays and next-generation sequencing data</li>
</ul>
</div>
<div id="specific-objectives" class="section level1">
<h1 class="hasAnchor">
<a href="#specific-objectives" class="anchor"></a>Specific objectives</h1>
<ul>
<li>Use <code>GRangesList</code> and <code>RaggedExperiment</code> to represent, access, and manipulate CNV data</li>
<li>Identify recurrent CNV regions in a population, including density trimming, reciprocal overlap, and recurrence significance estimation</li>
<li>Use the <a href="http://bioconductor.org/packages/regioneR">regioneR</a> package to assess the significance of overlaps between CNVs and functional genomic regions such as genes, promoters, and enhancers.</li>
<li>Carry out eQTL analysis for CNV and RNA-seq data using <code>GenomicRanges</code> / <code>RaggedExperiment</code> / <code>CNVRanger</code> architecture</li>
<li>Carry out a GWAS analysis for CNV and quantitative phenotype data</li>
<li>Perform estimation of tumor purity and ploidy from absolute CN analysis with <a href="http://bioconductor.org/packages/PureCN">PureCN</a>
</li>
</ul>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ludwig Geistlinger <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
