<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • CNVWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="CNVWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CNVWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Geistlinger_CNVWorkshop.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/waldronlab/CNVWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Geistlinger_CNVWorkshop_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="Geistlinger_CNVWorkshop_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="Geistlinger_CNVWorkshop_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      
      <div class="hidden name"><code>Geistlinger_CNVWorkshop.Rmd</code></div>

    </div>

    
    
<div id="copy-number-variation-analysis-with-bioconductor" class="section level1">
<h1 class="hasAnchor">
<a href="#copy-number-variation-analysis-with-bioconductor" class="anchor"></a>Copy number variation analysis with Bioconductor</h1>
<div id="workshop-information" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-information" class="anchor"></a>Workshop information</h2>
<div id="instructors-names-and-contact-information" class="section level3">
<h3 class="hasAnchor">
<a href="#instructors-names-and-contact-information" class="anchor"></a>Instructor(s) name(s) and contact information</h3>
<p>Ludwig Geistlinger, Marcel Ramos, Sehyun Oh, and Levi Waldron</p>
<p>CUNY School of Public Health 55 W 125th St, New York, NY 10027</p>
<p><a href="mailto:Ludwig_Geistlinger@hms.harvard.edu" class="email">Ludwig_Geistlinger@hms.harvard.edu</a>&lt;&gt; <a href="mailto:Marcel.Ramos@sph.cuny.edu" class="email">Marcel.Ramos@sph.cuny.edu</a>&lt;&gt; <a href="mailto:Sehyun.Oh@sph.cuny.edu" class="email">Sehyun.Oh@sph.cuny.edu</a>&lt;&gt; <a href="mailto:Levi.Waldron@sph.cuny.edu" class="email">Levi.Waldron@sph.cuny.edu</a></p>
</div>
<div id="workshop-description" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h3>
<p>This workshop gives an overview of Bioconductor solutions for the analysis of copy number variation (CNV) data. The workshop introduces Bioconductor core data structures for efficient representation, access, and manipulation of CNV data, and how to use these containers for structured downstream analysis of CNVs and integration with gene expression and quantitative phenotypes. Participants will be provided with code and hands-on practice for a comprehensive set of typical analysis steps including exploratory analysis, summarizing individual CNV calls across a population, overlap analysis with functional genomic regions and regulatory elements, expression quantitative trait loci (eQTL) analysis, and genome-wide association analysis (GWAS) with quantitative phenotypes. As an advanced application example, the workshop also introduces allele-specific absolute copy number analysis and how it is incorporated in genomic cancer analysis for the estimation of tumor characteristics such as tumor purity and ploidy.</p>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li><p>Basic knowledge of R syntax</p></li>
<li><p>Familiarity with the SummarizedExperiment class</p></li>
<li><p>Familiarity with the GenomicRanges class</p></li>
<li><p>Familiarity with high-throughput genomic assays such as microarrays and next-generation sequencing</p></li>
<li><p>Familiarity with the biological definition of single nucleotide polymorphism (SNP) and copy number variation (CNV)</p></li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>Execution of example code and hands-on practice</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li><a href="http://bioconductor.org/packages/RaggedExperiment">RaggedExperiment</a></li>
<li><a href="http://bioconductor.org/packages/CNVRanger">CNVRanger</a></li>
<li><a href="http://bioconductor.org/packages/regioneR">regioneR</a></li>
<li><a href="http://bioconductor.org/packages/PureCN">PureCN</a></li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Data representation and manipulation</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Integrative downstream analysis (eQTL, GWAS, …)</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Allele-specific CN analysis in cancer</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning Goals</h3>
<ul>
<li>get familiar with elementary concepts of CNV analysis</li>
<li>learn how to efficiently represent, access, and manipulate CNV data in Bioconductor data structures</li>
<li>get familiar with different strategies for summarizing individual CNV calls across a population</li>
<li>learn how to assess the significance of overlaps between CNVs and functional genomic regions</li>
<li>learn how carry out association analysis with gene expression and quantitative phenotypes</li>
<li>get familiar with allele-specific absolute CN analysis of genomic cancer data</li>
</ul>
</div>
<div id="specific-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#specific-objectives" class="anchor"></a>Specific objectives</h3>
<ul>
<li>understand how CNVs can be experimentally detected and computationally inferred from SNP arrays and next-generation sequencing data</li>
<li>learn how to use <code>GRangesList</code> and <code>RaggedExperiment</code> to represent, access, and manipulate CNV data</li>
<li>understand different strategies for finding recurrent CNV regions in a population, including density trimming, reciprocal overlap, and recurrence significance estimation</li>
<li>learn how to use the <a href="http://bioconductor.org/packages/regioneR">regioneR</a> package to assess the significance of overlaps between CNVs and functional genomic regions such as genes, promoters, and enhancers.</li>
<li>learn how to carry out eQTL analysis for CNV and RNA-seq data</li>
<li>learn how to carry out a GWAS analysis for CNV and quantitative phenotype data</li>
<li>learn how to estimate tumor purity and ploidy from absolute CN analysis with <a href="http://bioconductor.org/packages/PureCN">PureCN</a>
</li>
</ul>
</div>
</div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">CNVRanger</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">regioneR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Btaurus.UCSC.bosTau6.masked</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">curatedTCGAData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TCGAutils</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RaggedExperiment</span><span class="op">)</span></code></pre></div>
<p>Copy number variation (CNV) is a frequently observed deviation from the diploid state due to duplication or deletion of genomic regions. CNVs can be experimentally detected based on comparative genomic hybridization, and computationally inferred from SNP-arrays or next-generation sequencing data. These technologies for CNV detection have in common that they report, for each sample under study, genomic regions that are duplicated or deleted with respect to a reference. Such regions are denoted as <em>CNV calls</em> in the following and will be considered the starting point for analysis.</p>
<p>Relevant literature:</p>
<p>da Silva V, Ramos M, Groenen M, Crooijmans R, Johansson A, Regitano L, Coutinho L, Zimmer R, Waldron L, Geistlinger L. <a href="https://doi.org/10.1093/bioinformatics/btz632">CNVRanger: association analysis of CNVs with gene expression and quantitative phenotypes</a>. <em>Bioinformatics</em>, 36(3):972-3, 2020.</p>
<p>Oh S, Geistlinger L, Ramos M, Morgan M, Waldron L, Riester M. <a href="https://doi.org/10.1200/CCI.19.00130">Reliable analysis of clinical tumor-only whole-exome sequencing data</a>. <em>JCO Clin Cancer Inform</em>, 4:321-35, 2020.</p>
<p>Geistlinger L, Oh S, Ramos M, Schiffer L, LaRue R, Henzler C, Munro S, Daughters C, Nelson A, Winterhoff B, Chang Z, Talukdar S, Shetty M, Mullaney S, Morgan M, Parmigiani G, Birrer M, Qin LX, Riester M, Starr T, Waldron L. <a href="https://doi.org/10.1101/554394">Multi-omic analysis of subtype evolution and heterogeneity in high-grade serous ovarian carcinoma</a>. <em>Cancer Research</em>, 2020.</p>
</div>
<div id="representation-and-manipulation-of-cnv-data-with-raggedexperiment" class="section level2">
<h2 class="hasAnchor">
<a href="#representation-and-manipulation-of-cnv-data-with-raggedexperiment" class="anchor"></a>Representation and manipulation of CNV data with RaggedExperiment</h2>
<p><code>RaggedExperiment</code> is a flexible data representation for segmented copy number, somatic mutations such as represented in <code>.vcf</code> files, and other ragged array schema for genomic location data. Like the <code>GRangesList</code> class from <code>GenomicRanges</code>, <code>RaggedExperiment</code> can be used to represent <em>differing</em> genomic ranges on each of a set of samples. In fact, <code>RaggedExperiment</code> contains a <code>GRangesList</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">showClass</span><span class="op">(</span><span class="st">"RaggedExperiment"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "RaggedExperiment" [package "RaggedExperiment"]
## 
## Slots:
##                                                       
## Name:       assays      rowidx      colidx    metadata
## Class: GRangesList     integer     integer        list
## 
## Extends: "Annotated"</code></pre>
<div id="importing-ranged-data-with-genomicranges" class="section level3">
<h3 class="hasAnchor">
<a href="#importing-ranged-data-with-genomicranges" class="anchor"></a>Importing ranged data with <code>GenomicRanges</code>
</h3>
<p>When importing data from a file, you can represent an imported table (usually <code>data.frame</code>) with genomic information using helper functions in <code>GenomicRanges</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html">makeGRangesFromDataFrame</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span>
<span class="fu">makeGRangesListFromData</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></code></pre></div>
</div>
<div id="constructing-a-raggedexperiment-object" class="section level3">
<h3 class="hasAnchor">
<a href="#constructing-a-raggedexperiment-object" class="anchor"></a>Constructing a <code>RaggedExperiment</code> object</h3>
<p>We start with a toy example of two <code>GRanges</code> objects, providing ranges on two chromosomes in two samples:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="st">"chr1:1-10:-"</span>, B <span class="op">=</span> <span class="st">"chr1:8-14:+"</span>, C <span class="op">=</span> <span class="st">"chr1:15-18:+"</span><span class="op">)</span>,
    score <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"germline"</span>, <span class="st">"somatic"</span>, <span class="st">"germline"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sample2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>D <span class="op">=</span> <span class="st">"chr1:1-10:-"</span>, E <span class="op">=</span> <span class="st">"chr1:11-18:+"</span><span class="op">)</span>,
    score <span class="op">=</span> <span class="fl">11</span><span class="op">:</span><span class="fl">12</span>, type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"germline"</span>, <span class="st">"somatic"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Include column data <code>colData</code> to describe the samples:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">colDat</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span>id<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, status <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"case"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The <code>RaggedExperiment</code> can be constructed from individual <code>Granges</code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">ragexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/RaggedExperiment-class.html">RaggedExperiment</a></span><span class="op">(</span>
    sample1 <span class="op">=</span> <span class="va">sample1</span>,
    sample2 <span class="op">=</span> <span class="va">sample2</span>,
    colData <span class="op">=</span> <span class="va">colDat</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## class: RaggedExperiment 
## dim: 5 2 
## assays(2): score type
## rownames(5): A B C D E
## colnames(2): sample1 sample2
## colData names(2): id status</code></pre>
<p>Or from a <code>GRangesList</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html">GRangesList</a></span><span class="op">(</span>sample1<span class="op">=</span><span class="va">sample1</span>, sample2<span class="op">=</span><span class="va">sample2</span><span class="op">)</span>
<span class="va">ragexp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/RaggedExperiment-class.html">RaggedExperiment</a></span><span class="op">(</span><span class="va">grl</span>, colData <span class="op">=</span> <span class="va">colDat</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span><span class="op">(</span><span class="va">ragexp</span>, <span class="va">ragexp2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Note that the original ranges are is represented as the <code>rowRanges</code> of the <code>RaggedExperiment</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span><span class="op">(</span><span class="va">ragexp</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 5 ranges and 0 metadata columns:
##     seqnames    ranges strand
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   A     chr1      1-10      -
##   B     chr1      8-14      +
##   C     chr1     15-18      +
##   D     chr1      1-10      -
##   E     chr1     11-18      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="assay-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#assay-functions" class="anchor"></a><code>*Assay</code> functions</h3>
<p><code>RaggedExperiment</code> provides a flexible set of _*Assay_ methods to support transformation of data to matrix format with varying row dimensions.</p>
<div class="figure">
<img src="RaggedExperiment.png" alt="RaggedExperiment object schematic. Rows and columns represent genomic ranges and samples, respectively. Assay operations can be performed with (from left to right) compactAssay, qreduceAssay, and sparseAssay." width="\maxwidth"><p class="caption">
RaggedExperiment object schematic. Rows and columns represent genomic ranges and samples, respectively. Assay operations can be performed with (from left to right) compactAssay, qreduceAssay, and sparseAssay.
</p>
</div>
<p>The four main <em>Assay</em> functions for converting to matrix are:</p>
<ul>
<li>
<a href="#sparseassay">sparseAssay</a>: leave ranges exactly as-is</li>
<li>
<a href="#compactassay">compactAssay</a>: combine identical ranges</li>
<li>
<a href="#disjoinassay">disjoinAssay</a>: disjoin ranges that overlap across samples</li>
<li>
<a href="#qreduceassay">qreduceAssay</a>: find overlaps with provided “query” ranges</li>
</ul>
<p>These each have a corresponding function for conversion to <a href="#Conversion%20to%20RangedSummarizedExperiment">RangedSummarizedExperiment</a>.</p>
<div id="sparseassay" class="section level4">
<h4 class="hasAnchor">
<a href="#sparseassay" class="anchor"></a>sparseAssay</h4>
<p>The most straightforward matrix representation of a <code>RaggedExperiment</code> will return a matrix with the number of rows equal to the total number of ranges defined across all samples. <em>i.e.</em> the 5 rows of the <code>sparseAssay</code> result:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/assay-functions.html">sparseAssay</a></span><span class="op">(</span><span class="va">ragexp</span><span class="op">)</span></code></pre></div>
<pre><code>##   sample1 sample2
## A       3      NA
## B       4      NA
## C       5      NA
## D      NA      11
## E      NA      12</code></pre>
<p>correspond to the ranges of the unlisted <code>GRangesList</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 5 ranges and 2 metadata columns:
##             seqnames    ranges strand |     score        type
##                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   sample1.A     chr1      1-10      - |         3    germline
##   sample1.B     chr1      8-14      + |         4     somatic
##   sample1.C     chr1     15-18      + |         5    germline
##   sample2.D     chr1      1-10      - |        11    germline
##   sample2.E     chr1     11-18      + |        12     somatic
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>The rownames of the <code>sparseAssay</code> result are equal to the names of the <code>GRanges</code> elements. The values in the matrix returned by <code>sparseAssay</code> correspond to the first columns of the <code>mcols</code> of each <code>GRangesList</code> element, in this case the “score” column.</p>
<p>Note, this is the default <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay()</a></code> method of <code>RaggedExperiment</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">ragexp</span>, <span class="st">"score"</span><span class="op">)</span></code></pre></div>
<pre><code>##   sample1 sample2
## A       3      NA
## B       4      NA
## C       5      NA
## D      NA      11
## E      NA      12</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">ragexp</span>, <span class="st">"type"</span><span class="op">)</span></code></pre></div>
<pre><code>##   sample1    sample2   
## A "germline" NA        
## B "somatic"  NA        
## C "germline" NA        
## D NA         "germline"
## E NA         "somatic"</code></pre>
</div>
<div id="compactassay" class="section level4">
<h4 class="hasAnchor">
<a href="#compactassay" class="anchor"></a>compactAssay</h4>
<p>The dimensions of the <code>compactAssay</code> result differ from that of the <code>sparseAssay</code> result only if there are identical ranges in different samples. Identical ranges are placed in the same row in the output. Ranges with any difference in start, end, or strand, will be kept on different rows. Non-disjoint ranges are <strong>not</strong> collapsed.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/assay-functions.html">compactAssay</a></span><span class="op">(</span><span class="va">ragexp</span><span class="op">)</span></code></pre></div>
<pre><code>##              sample1 sample2
## chr1:8-14:+        4      NA
## chr1:11-18:+      NA      12
## chr1:15-18:+       5      NA
## chr1:1-10:-        3      11</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/assay-functions.html">compactAssay</a></span><span class="op">(</span><span class="va">ragexp</span>, <span class="st">"type"</span><span class="op">)</span></code></pre></div>
<pre><code>##              sample1    sample2   
## chr1:8-14:+  "somatic"  NA        
## chr1:11-18:+ NA         "somatic" 
## chr1:15-18:+ "germline" NA        
## chr1:1-10:-  "germline" "germline"</code></pre>
<p>Note that row names are constructed from the ranges, and the names of the <code>GRanges</code> vectors are lost, unlike in the <code>sparseAssay</code> result.</p>
</div>
<div id="disjoinassay" class="section level4">
<h4 class="hasAnchor">
<a href="#disjoinassay" class="anchor"></a>disjoinAssay</h4>
<p>This function is similar to <code>compactAssay</code> except the rows are <em>disjoint</em><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> ranges. Elements of the matrix are summarized by applying the <code>simplifyDisjoin</code> functional argument to assay values of overlapping ranges.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/assay-functions.html">disjoinAssay</a></span><span class="op">(</span><span class="va">ragexp</span>, simplifyDisjoin <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></code></pre></div>
<pre><code>##              sample1 sample2
## chr1:8-10:+        4      NA
## chr1:11-14:+       4      12
## chr1:15-18:+       5      12
## chr1:1-10:-        3      11</code></pre>
</div>
<div id="qreduceassay" class="section level4">
<h4 class="hasAnchor">
<a href="#qreduceassay" class="anchor"></a>qreduceAssay</h4>
<p>The <code>qreduceAssay</code> function is the most complicated but likely the most useful of the <code>RaggedExperiment</code> <em>Assay</em> functions. It requires you to provide a <code>query</code> argument that is a <code>GRanges</code> vector, and the rows of the resulting matrix correspond to the elements of this <code>GRanges</code>. The returned matrix will have dimensions <code><a href="https://rdrr.io/r/base/length.html">length(query)</a></code> by <code><a href="https://rdrr.io/r/base/nrow.html">ncol(x)</a></code>. Elements of the resulting matrix correspond to the overlap of the <em>i</em> th <code>query</code> range in the <em>j</em> th sample, summarized according to the <code>simplifyReduce</code> functional argument. This can be useful, for example, to calculate per-gene copy number or mutation status by providing the genomic ranges of every gene as the <code>query</code>.</p>
<p>The <code>simplifyReduce</code> argument in <code>qreduceAssay</code> allows the user to summarize overlapping regions with a custom method for the given “query” region of interest. We provide one for calculating a weighted average score per query range, where the weight is proportional to the overlap widths between overlapping ranges and a query range.</p>
<p><em>Note</em> that there are three arguments to this function. See the documentation for additional details.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">weightedmean</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">scores</span>, <span class="va">ranges</span>, <span class="va">qranges</span><span class="op">)</span>
<span class="op">{</span>
    <span class="va">isects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/setops-methods.html">pintersect</a></span><span class="op">(</span><span class="va">ranges</span>, <span class="va">qranges</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">scores</span> <span class="op">*</span> <span class="fu">width</span><span class="op">(</span><span class="va">isects</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">isects</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>The call to <code>qreduceAssay</code> calculates the overlaps between the ranges of each sample:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grl</span></code></pre></div>
<pre><code>## GRangesList object of length 2:
## $sample1
## GRanges object with 3 ranges and 2 metadata columns:
##     seqnames    ranges strand |     score        type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   A     chr1      1-10      - |         3    germline
##   B     chr1      8-14      + |         4     somatic
##   C     chr1     15-18      + |         5    germline
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $sample2
## GRanges object with 2 ranges and 2 metadata columns:
##     seqnames    ranges strand |     score        type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   D     chr1      1-10      - |        11    germline
##   E     chr1     11-18      + |        12     somatic
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>with the query ranges (an arbitrary set is defined here for demonstration): First create a demonstration “query” region of interest:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr1:1-14:-"</span>, <span class="st">"chr1:15-18:+"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-14      -
##   [2]     chr1     15-18      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>using the <code>simplifyReduce</code> function to resolve overlapping ranges and return a matrix with rows corresponding to the query:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/assay-functions.html">qreduceAssay</a></span><span class="op">(</span><span class="va">ragexp</span>, <span class="va">query</span>, simplifyReduce <span class="op">=</span> <span class="va">weightedmean</span><span class="op">)</span></code></pre></div>
<pre><code>##              sample1 sample2
## chr1:1-14:-        3      11
## chr1:15-18:+       5      12</code></pre>
</div>
</div>
<div id="conversion-to-rangedsummarizedexperiment" class="section level3">
<h3 class="hasAnchor">
<a href="#conversion-to-rangedsummarizedexperiment" class="anchor"></a>Conversion to RangedSummarizedExperiment</h3>
<p>These methods all have corresponding methods to return a <code>RangedSummarizedExperiment</code> and preserve the <code>colData</code>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/coerce-functions.html">sparseSummarizedExperiment</a></span><span class="op">(</span><span class="va">ragexp</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/coerce-functions.html">compactSummarizedExperiment</a></span><span class="op">(</span><span class="va">ragexp</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/coerce-functions.html">disjoinSummarizedExperiment</a></span><span class="op">(</span><span class="va">ragexp</span>, simplify <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/coerce-functions.html">qreduceSummarizedExperiment</a></span><span class="op">(</span><span class="va">ragexp</span>, query <span class="op">=</span> <span class="va">query</span>, simplify <span class="op">=</span> <span class="va">weightedmean</span><span class="op">)</span></code></pre></div>
<p>Please see the <code>RaggedExperiment</code> vignette for <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/RaggedExperiment/inst/doc/RaggedExperiment.html">more details</a>.</p>
</div>
</div>
<div id="integrative-downstream-analysis-of-cnvs-with-cnvranger" class="section level2">
<h2 class="hasAnchor">
<a href="#integrative-downstream-analysis-of-cnvs-with-cnvranger" class="anchor"></a>Integrative downstream analysis of CNVs with CNVRanger</h2>
<p><img src="CNVRanger.png" width="100%"></p>
<p><strong>(A)</strong> The <a href="http://bioconductor.org/packages/CNVRanger">CNVRanger</a> package imports CNV calls from a simple file format into <code>R</code>, and stores them in dedicated Bioconductor data structures, and <strong>(B)</strong> implements three frequently used approaches for summarizing CNV calls across a population: <em>(i)</em> the <a href="http://www.ircgp.com/CNVRuler">CNVRuler</a> that trims region margins based on regional density <a href="https://doi.org/10.1093/bioinformatics/bts239">Kim et al., 2012</a>, <em>(ii)</em> the reciprocal overlap procedure that requires sufficient mutual overlap between calls <a href="https://doi.org/10.1038/nature08516">Conrad et al., 2010</a>, and <em>(iii)</em> the <a href="http://www.broadinstitute.org/cancer/cga/gistic">GISTIC</a> procedure that identifies recurrent CNV regions <a href="https://doi.org/10.1073/pnas.0710052104">Beroukhim et al., 2007</a>. <strong>(C)</strong> <code>CNVRanger</code> builds on <a href="http://bioconductor.org/packages/regioneR">regioneR</a> for overlap analysis of CNVs with functional genomic regions, <strong>(D)</strong> implements RNA-seq expression Quantitative Trait Loci (eQTL) analysis for CNVs by interfacing with <a href="http://bioconductor.org/packages/edgeR">edgeR</a>, and <strong>(E)</strong> interfaces with <a href="http://zzz.bwh.harvard.edu/plink">PLINK</a> for traditional genome-wide association studies (GWAS) between CNVs and quantitative phenotypes.</p>
<p>The key parts of the functionality implemented in <code>CNVRanger</code> were developed, described, and applied in several previous studies:</p>
<ul>
<li><p>Genome-wide detection of CNVs and their association with meat tenderness in Nelore cattle <a href="https://doi.org/10.1371/journal.pone.0157711">da Silva et al., 2016</a></p></li>
<li><p>Widespread modulation of gene expression by copy number variation in skeletal muscle <a href="https://doi.org/10.1038/s41598-018-19782-4">Geistlinger et al., 2018</a></p></li>
<li><p>CNVs are associated with genomic architecture in a songbird <a href="https://doi.org/10.1186/s12864-018-4577-1">da Silva et al., 2018</a></p></li>
</ul>
<div id="reading-and-accessing-cnv-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-and-accessing-cnv-data" class="anchor"></a>Reading and accessing CNV data</h3>
<p><img src="Input.png" width="100%"></p>
<p><code>CNVRanger</code> uses Bioconductor core data structures implemented in the <a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a> and <a href="http://bioconductor.org/packages/RaggedExperiment">RaggedExperiment</a> packages to represent, access, and manipulate CNV data.</p>
<p>We start by loading the package.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">CNVRanger</span><span class="op">)</span></code></pre></div>
<div id="input-data-format" class="section level4">
<h4 class="hasAnchor">
<a href="#input-data-format" class="anchor"></a>Input data format</h4>
<p><code>CNVRanger</code> reads CNV calls from a simple file format, providing at least chromosome, start position, end position, sample ID, and integer copy number for each call.</p>
<p>For demonstration, we consider CNV calls as obtained with <a href="http://penncnv.openbioinformatics.org">PennCNV</a> from SNP-chip data in a Brazilian cattle breed (<a href="https://doi.org/10.1371/journal.pone.0157711">da Silva et al., 2016</a>).</p>
<p>Here, we use a data subset and only consider CNV calls on chromosome 1 and 2, for which there are roughly 3000 CNV calls as obtained for 711 samples. We use <code>read.csv</code> to read comma-separated values, but we could use a different function if the data were provided with a different delimiter (for example, <code>read.delim</code> for tab-separated values).</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"CNVRanger"</span><span class="op">)</span>
<span class="va">call.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">data.dir</span>, <span class="st">"Silva16_PONE_CNV_calls.csv"</span><span class="op">)</span>
<span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">call.file</span>, as.is<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 3000</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></code></pre></div>
<pre><code>##    chr start    end    NE_id state
## 1 chr1 16947  45013 NE001423     3
## 2 chr1 36337  67130 NE001426     3
## 3 chr1 16947  36337 NE001428     3
## 4 chr1 36337 105963 NE001519     3
## 5 chr1 36337  83412 NE001534     3
## 6 chr1 36337  83412 NE001648     3</code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">calls</span><span class="op">[</span>,<span class="st">"NE_id"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 711</code></pre>
<p>We observe that this example dataset satisfies the basic five-column input format required by <code>CNVRanger</code>.</p>
<p>The last column contains the integer copy number state for each call, encoded as</p>
<ul>
<li>
<code>0</code>: homozygous deletion (2-copy loss)</li>
<li>
<code>1</code>: heterozygous deletion (1-copy loss)</li>
<li>
<code>2</code>: normal diploid state</li>
<li>
<code>3</code>: 1-copy gain</li>
<li>
<code>4</code>: amplification (&gt;= 2-copy gain)</li>
</ul>
</div>
<div id="representation-as-a-grangeslist" class="section level4">
<h4 class="hasAnchor">
<a href="#representation-as-a-grangeslist" class="anchor"></a>Representation as a <code>GRangesList</code>
</h4>
<p>Once read into an R <code>data.frame</code>, we group the calls by sample ID and convert them to a <code>GRangesList</code>. Each element of the list corresponds to a sample, and contains the genomic coordinates of the CNV calls for this sample (along with the copy number state in the <code>state</code> metadata column).</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesListFromDataFrame.html">makeGRangesListFromDataFrame</a></span><span class="op">(</span><span class="va">calls</span>, 
    split.field <span class="op">=</span> <span class="st">"NE_id"</span>, keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">grl</span></code></pre></div>
<pre><code>## GRangesList object of length 711:
## $NE001357
## GRanges object with 5 ranges and 1 metadata column:
##       seqnames            ranges strand |     state
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]     chr1   4569526-4577608      * |         3
##   [2]     chr1 15984544-15996851      * |         1
##   [3]     chr1 38306432-38330161      * |         3
##   [4]     chr1 93730576-93819471      * |         0
##   [5]     chr2 40529044-40540747      * |         3
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $NE001358
## GRanges object with 1 range and 1 metadata column:
##       seqnames              ranges strand |     state
##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]     chr1 105042452-105233446      * |         1
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $NE001359
## GRanges object with 4 ranges and 1 metadata column:
##       seqnames            ranges strand |     state
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]     chr1   4569526-4577608      * |         3
##   [2]     chr1 31686841-31695808      * |         0
##   [3]     chr1 93730576-93819471      * |         0
##   [4]     chr2   2527718-2535261      * |         0
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## ...
## &lt;708 more elements&gt;</code></pre>
<p>The advantage of representing the CNV calls as a <code>GRangesList</code> is that it allows to leverage the comprehensive set of operations on genomic regions implemented in the <code>GenomicRanges</code> package - for instance, sorting of the calls according to their genomic coordinates.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span>
<span class="va">grl</span></code></pre></div>
<pre><code>## GRangesList object of length 711:
## $NE001357
## GRanges object with 5 ranges and 1 metadata column:
##       seqnames            ranges strand |     state
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]     chr1   4569526-4577608      * |         3
##   [2]     chr1 15984544-15996851      * |         1
##   [3]     chr1 38306432-38330161      * |         3
##   [4]     chr1 93730576-93819471      * |         0
##   [5]     chr2 40529044-40540747      * |         3
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $NE001358
## GRanges object with 1 range and 1 metadata column:
##       seqnames              ranges strand |     state
##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]     chr1 105042452-105233446      * |         1
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $NE001359
## GRanges object with 4 ranges and 1 metadata column:
##       seqnames            ranges strand |     state
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   [1]     chr1   4569526-4577608      * |         3
##   [2]     chr1 31686841-31695808      * |         0
##   [3]     chr1 93730576-93819471      * |         0
##   [4]     chr2   2527718-2535261      * |         0
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## ...
## &lt;708 more elements&gt;</code></pre>
</div>
<div id="representation-as-a-raggedexperiment" class="section level4">
<h4 class="hasAnchor">
<a href="#representation-as-a-raggedexperiment" class="anchor"></a>Representation as a <code>RaggedExperiment</code>
</h4>
<p>An alternative matrix-like representation of the CNV calls can be obtained with the <code>RaggedExperiment</code> data class. It resembles in many aspects the <a href="http://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a> data class for storing gene expression data as e.g. obtained with RNA-seq.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ra</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RaggedExperiment/man/RaggedExperiment-class.html">RaggedExperiment</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span>
<span class="va">ra</span></code></pre></div>
<pre><code>## class: RaggedExperiment 
## dim: 3000 711 
## assays(1): state
## rownames: NULL
## colnames(711): NE001357 NE001358 ... NE003967 NE003968
## colData names(0):</code></pre>
<p>As apparent from the <code>dim</code> slot of the object, it stores the CNV calls in the rows and the samples in the columns. Note that the CN state is now represented as an assay matrix which can be easily accessed and subsetted.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">ra</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##                        NE001357 NE001358 NE001359 NE001360 NE001361
## chr1:4569526-4577608          3       NA       NA       NA       NA
## chr1:15984544-15996851        1       NA       NA       NA       NA
## chr1:38306432-38330161        3       NA       NA       NA       NA
## chr1:93730576-93819471        0       NA       NA       NA       NA
## chr2:40529044-40540747        3       NA       NA       NA       NA</code></pre>
<p>As with <code>SummarizedExperiment</code> objects, additional information for the samples are annotated in the <code>colData</code> slot. For example, we annotate the steer weight and its feed conversion ratio (FCR) using simulated data. Feed conversion ratio is the ratio of dry matter intake to live-weight gain. A typical range of feed conversion ratios is 4.5-7.5 with a lower number being more desirable as it would indicate that a steer required less feed per pound of gain.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">ra</span><span class="op">)</span>, mean<span class="op">=</span><span class="fl">1100</span>, sd<span class="op">=</span><span class="fl">100</span><span class="op">)</span>
<span class="va">fcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">ra</span><span class="op">)</span>, mean<span class="op">=</span><span class="fl">6</span>, sd<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">ra</span><span class="op">)</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">weight</span>, digits<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">ra</span><span class="op">)</span><span class="op">$</span><span class="va">fcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">fcr</span>, digits<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">ra</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 711 rows and 2 columns
##             weight       fcr
##          &lt;numeric&gt; &lt;numeric&gt;
## NE001357   1273.89      6.33
## NE001358    948.57      5.64
## NE001359   1223.83      5.08
## NE001360   1166.69      8.42
## NE001361   1004.45      6.15
## ...            ...       ...
## NE003962   1088.90      7.25
## NE003963   1070.24      5.35
## NE003966   1270.36      6.12
## NE003967   1061.96      6.88
## NE003968   1116.48      4.60</code></pre>
</div>
</div>
<div id="summarizing-individual-cnv-calls-across-a-population" class="section level3">
<h3 class="hasAnchor">
<a href="#summarizing-individual-cnv-calls-across-a-population" class="anchor"></a>Summarizing individual CNV calls across a population</h3>
<p><img src="Summarization.png" width="40%" style="float:right; padding:10px"></p>
<p>In CNV analysis, it is often of interest to summarize individual calls across the population, (i.e. to define CNV regions), for subsequent association analysis with expression and phenotype data. In the simplest case, this just merges overlapping individual calls into summarized regions. However, this typically inflates CNV region size, and more appropriate approaches have been developed for this purpose.</p>
<div id="trimming-low-density-areas" class="section level4">
<h4 class="hasAnchor">
<a href="#trimming-low-density-areas" class="anchor"></a>Trimming low-density areas</h4>
<p>Here, we use the approach from <a href="http://www.ircgp.com/CNVRuler">CNVRuler</a> to summarize CNV calls to CNV regions (see <a href="https://academic.oup.com/view-large/figure/83392426/bts239f1.jpeg">Figure 1</a> in <a href="https://doi.org/10.1093/bioinformatics/bts239">Kim et al., 2012</a> for an illustration of the approach). This trims low-density areas as defined by the <code>density</code> argument, which is set here to &lt;10% of the number of calls within a summarized region.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnvrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CNVRanger/man/populationRanges.html">populationRanges</a></span><span class="op">(</span><span class="va">grl</span>, density <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>
<span class="va">cnvrs</span></code></pre></div>
<pre><code>## GRanges object with 303 ranges and 2 metadata columns:
##         seqnames              ranges strand |      freq        type
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;character&gt;
##     [1]     chr1        16947-111645      * |       103        gain
##     [2]     chr1     1419261-1630187      * |        18        gain
##     [3]     chr1     1896112-2004603      * |       218        gain
##     [4]     chr1     4139727-4203274      * |         1        gain
##     [5]     chr1     4554832-4577608      * |        23        gain
##     ...      ...                 ...    ... .       ...         ...
##   [299]     chr2 136310067-136322489      * |         2        loss
##   [300]     chr2 136375337-136386940      * |         1        loss
##   [301]     chr2 136455546-136466040      * |         1        loss
##   [302]     chr2 136749793-136802493      * |        39        both
##   [303]     chr2 139194749-139665914      * |        58        both
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Note that CNV frequency (number of samples overlapping each region) and CNV type (gain, loss, or both) have also been annotated in the columns <code>freq</code> and <code>type</code>, respectively.</p>
</div>
<div id="identifying-recurrent-regions" class="section level4">
<h4 class="hasAnchor">
<a href="#identifying-recurrent-regions" class="anchor"></a>Identifying recurrent regions</h4>
<p>In particular in cancer, it is important to distinguish driver from passenger mutations, i.e. to distinguish meaningful events from random background aberrations. The <a href="http://www.broadinstitute.org/cancer/cga/gistic">GISTIC</a> method identifies those regions of the genome that are aberrant more often than would be expected by chance, with greater weight given to high amplitude events (high-level copy-number gains or homozygous deletions) that are less likely to represent random aberrations (<a href="https://doi.org/10.1073/pnas.0710052104">Beroukhim et al., 2007</a>).</p>
<p>By setting <code>est.recur=TRUE</code>, we deploy a <code>GISTIC</code>-like significance estimation</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cnvrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CNVRanger/man/populationRanges.html">populationRanges</a></span><span class="op">(</span><span class="va">grl</span>, density <span class="op">=</span> <span class="fl">0.1</span>, est.recur <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">cnvrs</span></code></pre></div>
<pre><code>## GRanges object with 303 ranges and 3 metadata columns:
##         seqnames              ranges strand |      freq        type     pvalue
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;character&gt;  &lt;numeric&gt;
##     [1]     chr1        16947-111645      * |       103        gain 0.00980392
##     [2]     chr1     1419261-1630187      * |        18        gain 0.10784314
##     [3]     chr1     1896112-2004603      * |       218        gain 0.00000000
##     [4]     chr1     4139727-4203274      * |         1        gain 0.55882353
##     [5]     chr1     4554832-4577608      * |        23        gain 0.08823529
##     ...      ...                 ...    ... .       ...         ...        ...
##   [299]     chr2 136310067-136322489      * |         2        loss  0.2361111
##   [300]     chr2 136375337-136386940      * |         1        loss  0.4212963
##   [301]     chr2 136455546-136466040      * |         1        loss  0.4212963
##   [302]     chr2 136749793-136802493      * |        39        both  0.0588235
##   [303]     chr2 139194749-139665914      * |        58        both  0.0392157
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>and filter for recurrent CNVs that exceed a significance threshold of 0.05.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">cnvrs</span>, <span class="va">pvalue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 17 ranges and 3 metadata columns:
##        seqnames              ranges strand |      freq        type     pvalue
##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;character&gt;  &lt;numeric&gt;
##    [1]     chr1        16947-111645      * |       103        gain 0.00980392
##    [2]     chr1     1896112-2004603      * |       218        gain 0.00000000
##    [3]     chr1   15984544-15996851      * |       116        loss 0.01851852
##    [4]     chr1   31686841-31695808      * |       274        loss 0.00462963
##    [5]     chr1   69205418-69219486      * |        46        loss 0.04166667
##    ...      ...                 ...    ... .       ...         ...        ...
##   [13]     chr2   97882695-97896935      * |        80        loss  0.0231481
##   [14]     chr2 124330343-124398570      * |        39        loss  0.0462963
##   [15]     chr2 135096060-135271140      * |        84        gain  0.0196078
##   [16]     chr2 135290754-135553033      * |        83        gain  0.0294118
##   [17]     chr2 139194749-139665914      * |        58        both  0.0392157
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>We can illustrate the landscape of recurrent CNV regions using the function <code>plotRecurrentRegions</code>. We therefore provide the summarized CNV regions, a valid <a href="https://genome.ucsc.edu/cgi-bin/hgGateway">UCSC</a> genome assembly, and a chromosome of interest.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/CNVRanger/man/plotRecurrentRegions.html">plotRecurrentRegions</a></span><span class="op">(</span><span class="va">cnvrs</span>, genome <span class="op">=</span> <span class="st">"bosTau6"</span>, chr <span class="op">=</span> <span class="st">"chr1"</span><span class="op">)</span></code></pre></div>
<p><img src="Geistlinger_CNVWorkshop_files/figure-html/recurrViz-1.png" width="100%"></p>
<p>The function plots (from top to bottom): (i) an ideogram of the chromosome (note that staining bands are not available for <code>bosTau6</code>), (ii) a genome axis indicating the chromosomal position, (iii) a bar plot showing for each CNV region the number of samples with a CNV call in that region, and (iv) an annotation track that indicates whether this is a <em>recurrent</em> region according to a significance threshold (an argument to the function, default: 0.05).</p>
</div>
</div>
<div id="overlap-analysis-of-cnvs-with-functional-genomic-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#overlap-analysis-of-cnvs-with-functional-genomic-regions" class="anchor"></a>Overlap analysis of CNVs with functional genomic regions</h3>
<p><img src="Overlap.png" width="40%" style="float:right; padding:10px"></p>
<p>Once individual CNV calls have been summarized across the population, it is typically of interest whether the resulting CNV regions overlap with functional genomic regions such as genes, promoters, or enhancers.</p>
<p>To obtain the location of protein-coding genes, we query available <em>Bos taurus</em> annotation from Ensembl</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span>
<span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu">AnnotationHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## using temporary cache /tmp/RtmpLQImfA/BiocFileCache</code></pre>
<pre><code>## snapshotDate(): 2020-10-27</code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ahDb</span> <span class="op">&lt;-</span> <span class="fu">AnnotationHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span><span class="op">(</span><span class="va">ah</span>, pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Bos taurus"</span>, <span class="st">"EnsDb"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ahDb</span></code></pre></div>
<pre><code>## AnnotationHub with 15 records
## # snapshotDate(): 2020-10-27
## # $dataprovider: Ensembl
## # $species: Bos taurus
## # $rdataclass: EnsDb
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
## #   rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[["AH53189"]]' 
## 
##             title                           
##   AH53189 | Ensembl 87 EnsDb for Bos Taurus 
##   AH53693 | Ensembl 88 EnsDb for Bos Taurus 
##   AH56658 | Ensembl 89 EnsDb for Bos Taurus 
##   AH57731 | Ensembl 90 EnsDb for Bos Taurus 
##   AH60745 | Ensembl 91 EnsDb for Bos Taurus 
##   ...       ...                             
##   AH73832 | Ensembl 97 EnsDb for Bos taurus 
##   AH74957 | Ensembl 98 EnsDb for Bos taurus 
##   AH78724 | Ensembl 99 EnsDb for Bos taurus 
##   AH79622 | Ensembl 100 EnsDb for Bos taurus
##   AH83145 | Ensembl 101 EnsDb for Bos taurus</code></pre>
<p>and retrieve gene coordinates in the UMD3.1 assembly (Ensembl 92).</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ahEdb</span> <span class="op">&lt;-</span> <span class="va">ahDb</span><span class="op">[[</span><span class="st">"AH60948"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code>## downloading 1 resources</code></pre>
<pre><code>## retrieving 1 resource</code></pre>
<pre><code>## loading from cache</code></pre>
<pre><code>## require("ensembldb")</code></pre>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bt.genes</span> <span class="op">&lt;-</span> <span class="fu">ensembldb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ensembldb/man/EnsDb-exonsBy.html">genes</a></span><span class="op">(</span><span class="va">ahEdb</span><span class="op">)</span>
<span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html">seqlevelsStyle</a></span><span class="op">(</span><span class="va">bt.genes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"UCSC"</span>
<span class="va">bt.genes</span></code></pre></div>
<pre><code>## GRanges object with 24616 ranges and 8 metadata columns:
##                      seqnames              ranges strand |            gene_id
##                         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt;
##   ENSBTAG00000046619     chr1         19774-19899      - | ENSBTAG00000046619
##   ENSBTAG00000006858     chr1         34627-35558      + | ENSBTAG00000006858
##   ENSBTAG00000039257     chr1         69695-71121      - | ENSBTAG00000039257
##   ENSBTAG00000035349     chr1         83323-84281      - | ENSBTAG00000035349
##   ENSBTAG00000001753     chr1       124849-179713      - | ENSBTAG00000001753
##                  ...      ...                 ...    ... .                ...
##   ENSBTAG00000025951     chrX 148526584-148535857      + | ENSBTAG00000025951
##   ENSBTAG00000029592     chrX 148538917-148539037      - | ENSBTAG00000029592
##   ENSBTAG00000016989     chrX 148576705-148582356      - | ENSBTAG00000016989
##   ENSBTAG00000025952     chrX 148774930-148780901      - | ENSBTAG00000025952
##   ENSBTAG00000047839     chrX 148804071-148805135      + | ENSBTAG00000047839
##                        gene_name   gene_biotype seq_coord_system
##                      &lt;character&gt;    &lt;character&gt;      &lt;character&gt;
##   ENSBTAG00000046619     RF00001           rRNA       chromosome
##   ENSBTAG00000006858                 pseudogene       chromosome
##   ENSBTAG00000039257             protein_coding       chromosome
##   ENSBTAG00000035349                 pseudogene       chromosome
##   ENSBTAG00000001753             protein_coding       chromosome
##                  ...         ...            ...              ...
##   ENSBTAG00000025951             protein_coding       chromosome
##   ENSBTAG00000029592     RF00001           rRNA       chromosome
##   ENSBTAG00000016989             protein_coding       chromosome
##   ENSBTAG00000025952             protein_coding       chromosome
##   ENSBTAG00000047839       P2RY8 protein_coding       chromosome
##                                 description      gene_id_version      symbol
##                                 &lt;character&gt;          &lt;character&gt; &lt;character&gt;
##   ENSBTAG00000046619                   NULL ENSBTAG00000046619.1     RF00001
##   ENSBTAG00000006858                   NULL ENSBTAG00000006858.5            
##   ENSBTAG00000039257                   NULL ENSBTAG00000039257.2            
##   ENSBTAG00000035349                   NULL ENSBTAG00000035349.3            
##   ENSBTAG00000001753                   NULL ENSBTAG00000001753.4            
##                  ...                    ...                  ...         ...
##   ENSBTAG00000025951                   NULL ENSBTAG00000025951.4            
##   ENSBTAG00000029592                   NULL ENSBTAG00000029592.1     RF00001
##   ENSBTAG00000016989                   NULL ENSBTAG00000016989.5            
##   ENSBTAG00000025952                   NULL ENSBTAG00000025952.3            
##   ENSBTAG00000047839 P2Y receptor family .. ENSBTAG00000047839.1       P2RY8
##                       entrezid
##                         &lt;list&gt;
##   ENSBTAG00000046619      &lt;NA&gt;
##   ENSBTAG00000006858      &lt;NA&gt;
##   ENSBTAG00000039257      &lt;NA&gt;
##   ENSBTAG00000035349      &lt;NA&gt;
##   ENSBTAG00000001753    507243
##                  ...       ...
##   ENSBTAG00000025951      &lt;NA&gt;
##   ENSBTAG00000029592      &lt;NA&gt;
##   ENSBTAG00000016989      &lt;NA&gt;
##   ENSBTAG00000025952    785083
##   ENSBTAG00000047839 100299937
##   -------
##   seqinfo: 48 sequences from UMD3.1 genome</code></pre>
<p>To speed up the example, we restrict analysis to chromosomes 1 and 2.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sel.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">bt.genes</span>, <span class="va">seqnames</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="va">sel.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">sel.genes</span>, <span class="va">gene_biotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span><span class="op">)</span>
<span class="va">sel.cnvrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">cnvrs</span>, <span class="va">seqnames</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="finding-and-illustrating-overlaps" class="section level4">
<h4 class="hasAnchor">
<a href="#finding-and-illustrating-overlaps" class="anchor"></a>Finding and illustrating overlaps</h4>
<p>The <code>findOverlaps</code> function from the <code>GenomicRanges</code> package is a general function for finding overlaps between two sets of genomic regions. Here, we use the function to find protein-coding genes (our <code>query</code> region set) overlapping the summarized CNV regions (our <code>subject</code> region set).</p>
<p>Resulting overlaps are represented as a <code>Hits</code> object, from which overlapping query and subject regions can be obtained with dedicated accessor functions (named <code>queryHits</code> and <code>subjectHits</code>, respectively). Here, we use these functions to also annotate the CNV type (gain/loss) for genes overlapping with CNVs.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">olaps</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span><span class="va">sel.genes</span>, <span class="va">sel.cnvrs</span>, ignore.strand<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">qh</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Hits-class.html">queryHits</a></span><span class="op">(</span><span class="va">olaps</span><span class="op">)</span>
<span class="va">sh</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Hits-class.html">subjectHits</a></span><span class="op">(</span><span class="va">olaps</span><span class="op">)</span>
<span class="va">cgenes</span> <span class="op">&lt;-</span> <span class="va">sel.genes</span><span class="op">[</span><span class="va">qh</span><span class="op">]</span>
<span class="va">cgenes</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="va">sel.cnvrs</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="va">sh</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">cgenes</span>, select <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 123 ranges and 1 metadata column:
##                      seqnames              ranges strand |        type
##                         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   ENSBTAG00000039257     chr1         69695-71121      - |        gain
##   ENSBTAG00000021819     chr1     1467704-1496151      - |        gain
##   ENSBTAG00000019404     chr1     1563137-1591758      - |        gain
##   ENSBTAG00000015212     chr1     1593295-1627137      - |        gain
##   ENSBTAG00000000597     chr1   18058709-18207251      + |        loss
##                  ...      ...                 ...    ... .         ...
##   ENSBTAG00000003822     chr2 136193743-136239981      - |        loss
##   ENSBTAG00000013281     chr2 136276529-136314563      + |        loss
##   ENSBTAG00000009251     chr2 136317925-136337845      - |        loss
##   ENSBTAG00000008510     chr2 136362255-136444097      + |        loss
##   ENSBTAG00000014221     chr2 136457565-136461977      + |        loss
##   -------
##   seqinfo: 48 sequences from UMD3.1 genome</code></pre>
<p>It might also be of interest to illustrate the original CNV calls on overlapping genomic features (here: protein-coding genes). For this purpose, an <code>oncoPrint</code> plot provides a useful summary in a rectangular fashion (genes in the rows, samples in the columns). Stacked barplots on the top and the right of the plot display the number of altered genes per sample and the number of altered samples per gene, respectively.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/CNVRanger/man/cnvOncoPrint.html">cnvOncoPrint</a></span><span class="op">(</span><span class="va">grl</span>, <span class="va">cgenes</span><span class="op">)</span></code></pre></div>
<pre><code>## Following `at` are removed: GAIN2, because no color was defined for
## them.
## Following `at` are removed: GAIN2, because no color was defined for
## them.
## Following `at` are removed: GAIN2, because no color was defined for
## them.</code></pre>
<p><img src="Geistlinger_CNVWorkshop_files/figure-html/vizOlaps-1.png" width="100%"></p>
</div>
<div id="overlap-permutation-test" class="section level4">
<h4 class="hasAnchor">
<a href="#overlap-permutation-test" class="anchor"></a>Overlap permutation test</h4>
<p>As a certain amount of overlap can be expected just by chance, an assessment of statistical significance is needed to decide whether the observed overlap is greater (enrichment) or less (depletion) than expected by chance.</p>
<p>The <a href="http://bioconductor.org/packages/regioneR">regioneR</a> package implements a general framework for testing overlaps of genomic regions based on permutation sampling. This allows to repeatedly sample random regions from the genome, matching size and chromosomal distribution of the region set under study (here: the CNV regions). By recomputing the overlap with the functional features in each permutation, statistical significance of the observed overlap can be assessed.</p>
<p>We demonstrate in the following how this strategy can be used to assess the overlap between the detected CNV regions and protein-coding regions in the cattle genome. We expect to find a depletion as protein-coding regions are highly conserved and rarely subject to long-range structural variation such as CNV. Hence, is the overlap between CNVs and protein-coding genes less than expected by chance?</p>
<p>To answer this question, we apply an overlap permutation test with 100 permutations (<code>ntimes=100</code>), while maintaining chromosomal distribution of the CNV region set (<code>per.chromosome=TRUE</code>). Furthermore, we use the option <code>count.once=TRUE</code> to count an overlapping CNV region only once, even if it overlaps with 2 or more genes. We also allow random regions to be sampled from the entire genome (<code>mask=NA</code>), although in certain scenarios masking certain regions such as telomeres and centromeres is advisable. Also note that we use 100 permutations for demonstration only. To draw robust conclusions a minimum of 1000 permutations should be carried out.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">regioneR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Btaurus.UCSC.bosTau6.masked</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/regioneR/man/overlapPermTest.html">overlapPermTest</a></span><span class="op">(</span>A<span class="op">=</span><span class="va">sel.cnvrs</span>, B<span class="op">=</span><span class="va">sel.genes</span>, ntimes<span class="op">=</span><span class="fl">100</span>, 
    genome<span class="op">=</span><span class="st">"bosTau6"</span>, mask<span class="op">=</span><span class="cn">NA</span>, per.chromosome<span class="op">=</span><span class="cn">TRUE</span>, count.once<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span></code></pre></div>
<pre><code>## $numOverlaps
## P-value: 0.0297029702970297
## Z-score: -2.074
## Number of iterations: 100
## Alternative: less
## Evaluation of the original region set: 97
## Evaluation function: numOverlaps
## Randomization function: randomizeRegions
## 
## attr(,"class")
## [1] "permTestResultsList"</code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">permuted</span><span class="op">)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##    93.0   110.0   114.0   115.1   122.2   135.0</code></pre>
<p>The resulting permutation <em>p</em>-value indicates a significant depletion. Out of the 303 CNV regions, 97 overlap with at least one gene. In contrast, when repeatedly drawing random regions matching the CNV regions in size and chromosomal distribution, the mean number of overlapping regions across permutations was 115.1 <span class="math inline">\(\pm\)</span> 8.7.</p>
<p>This finding is consistent with our observations across the whole genome (<a href="https://doi.org/10.1371/journal.pone.0157711">da Silva et al., 2016</a>) and findings from the 1000 Genomes Poject (<a href="https://www.nature.com/articles/nature15394">Sudmant et al., 2015</a>).</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<p><img src="Geistlinger_CNVWorkshop_files/figure-html/vizPermTest-1.png" width="100%"></p>
<p>Note: the function <code><a href="https://rdrr.io/pkg/regioneR/man/permTest.html">regioneR::permTest</a></code> allows to incorporate user-defined functions for randomizing regions and evaluating additional measures of overlap such as total genomic size in bp.</p>
</div>
</div>
<div id="cnv-expression-association-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#cnv-expression-association-analysis" class="anchor"></a>CNV-expression association analysis</h3>
<p><img src="Expression.png" width="40%" style="float:right; padding:10px"></p>
<p>Studies of expression quantitative trait loci (eQTLs) aim at the discovery of genetic variants that explain variation in gene expression levels (<a href="https://www.ncbi.nlm.nih.gov/pubmed/23650636">Nica and Dermitzakis, 2013</a>). Mainly applied in the context of SNPs, the concept also naturally extends to the analysis of CNVs.</p>
<p>The <code>CNVRanger</code> package implements association testing between CNV regions and RNA-seq read counts using <code>edgeR</code>, which applies generalized linear models based on the negative-binomial distribution while incorporating normalization factors for different library sizes.</p>
<p>In the case of only one CN state deviating from 2n for a CNV region under investigation, this reduces to the classical 2-group comparison. For more than two states (e.g. 0n, 1n, 2n), edgeR’s ANOVA-like test is applied to test all deviating groups for significant expression differences relative to 2n.</p>
<div id="working-with-individual-cnv-and-rna-seq-assays" class="section level4">
<h4 class="hasAnchor">
<a href="#working-with-individual-cnv-and-rna-seq-assays" class="anchor"></a>Working with individual CNV and RNA-seq assays</h4>
<p>We demonstrate the functionality by loading RNA-seq read count data from skeletal muscle samples for 183 Nelore cattle steers, which we analyzed for CNV-expression effects as previously described (<a href="https://doi.org/10.1038/s41598-018-19782-4">Geistlinger et al., 2018</a>).</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rseq.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">data.dir</span>, <span class="st">"counts_cleaned.txt"</span><span class="op">)</span>
<span class="va">rcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">rseq.file</span>, row.names<span class="op">=</span><span class="fl">1</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">rcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">rcounts</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">rcounts</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 939 183</code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rcounts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code>##                    NE001407 NE001408 NE001424 NE001439 NE001448
## ENSBTAG00000000088       64       65      233      135      134
## ENSBTAG00000000160       20       28       50       13       18
## ENSBTAG00000000176      279      373      679      223      417
## ENSBTAG00000000201      252      271      544      155      334
## ENSBTAG00000000210      268      379      486      172      443</code></pre>
<p>For demonstration, we restrict analysis to the 939 genes on chromosome 1 and 2, and store the RNA-seq expression data in a <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span>
<span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment</a></span><span class="op">(</span>assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>rcounts<span class="op">=</span><span class="va">rcounts</span><span class="op">)</span>, 
                rowRanges<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html">granges</a></span><span class="op">(</span><span class="va">sel.genes</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">rcounts</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="va">rse</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 939 183 
## metadata(0):
## assays(1): rcounts
## rownames(939): ENSBTAG00000000088 ENSBTAG00000000160 ...
##   ENSBTAG00000048210 ENSBTAG00000048228
## rowData names(0):
## colnames(183): NE001407 NE001408 ... NE003840 NE003843
## colData names(0):</code></pre>
<p>Assuming distinct modes of action, effects observed in the CNV-expression analysis are typically divided into (i) local effects (<em>cis</em>), where expression changes coincide with CNVs in the respective genes, and (ii) distal effects (<em>trans</em>), where CNVs supposedly affect trans-acting regulators such as transcription factors.</p>
<p>However, due to power considerations and to avoid detection of spurious effects, stringent filtering of (i) not sufficiently expressed genes, and (ii) CNV regions with insufficient sample size in groups deviating from 2n, should be carried out when testing for distal effects. Local effects have a clear spatial indication and the number of genes locating in or close to a CNV region of interest is typically small; testing for differential expression between CN states is thus generally better powered for local effects and less stringent filter criteria can be applied.</p>
<p>In the following, we carry out CNV-expression association analysis by providing the CNV regions to test (<code>cnvrs</code>), the individual CNV calls (<code>grl</code>) to determine per-sample CN state in each CNV region, the RNA-seq read counts (<code>rse</code>), and the size of the genomic window around each CNV region (<code>window</code>). The <code>window</code> argument thereby determines which genes are considered for testing for each CNV region and is set here to 1 Mbp.</p>
<p>Further, use the <code>filter.by.expr</code> and <code>min.samples</code> arguments to exclude from the analysis (i) genes with very low read count across samples, and (ii) CNV regions with fewer than <code>min.samples</code> samples in a group deviating from 2n.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CNVRanger/man/cnvEQTL.html">cnvEQTL</a></span><span class="op">(</span><span class="va">cnvrs</span>, <span class="va">grl</span>, <span class="va">rse</span>, window<span class="op">=</span><span class="st">"1Mbp"</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## Restricting analysis to 179 intersecting samples</code></pre>
<pre><code>## Preprocessing RNA-seq data ...</code></pre>
<pre><code>## Summarizing per-sample CN state in each CNV region</code></pre>
<pre><code>## Excluding 286 cnvrs not satisfying min.samples threshold</code></pre>
<pre><code>## Analyzing 12 regions with &gt;=1 gene in the given window</code></pre>
<pre><code>## 1 of 12</code></pre>
<pre><code>## 2 of 12</code></pre>
<pre><code>## 3 of 12</code></pre>
<pre><code>## 4 of 12</code></pre>
<pre><code>## 5 of 12</code></pre>
<pre><code>## 6 of 12</code></pre>
<pre><code>## 7 of 12</code></pre>
<pre><code>## 8 of 12</code></pre>
<pre><code>## 9 of 12</code></pre>
<pre><code>## 10 of 12</code></pre>
<pre><code>## 11 of 12</code></pre>
<pre><code>## 12 of 12</code></pre>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span></code></pre></div>
<pre><code>## GRangesList object of length 12:
## $`chr1:16947-111645`
## GRanges object with 5 ranges and 5 metadata columns:
##                      seqnames         ranges strand | logFC.CN0 logFC.CN1
##                         &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;
##   ENSBTAG00000018278     chr1  922635-929992      + |        NA        NA
##   ENSBTAG00000021997     chr1 944294-1188287      - |        NA        NA
##   ENSBTAG00000020035     chr1  351708-362907      + |        NA        NA
##   ENSBTAG00000011528     chr1  463572-478996      - |        NA        NA
##   ENSBTAG00000012594     chr1  669920-733729      - |        NA        NA
##                        logFC.CN3     PValue AdjPValue
##                        &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
##   ENSBTAG00000018278  0.19485971 0.00731515  0.400665
##   ENSBTAG00000021997 -0.08124960 0.17961602  0.841565
##   ENSBTAG00000020035  0.07450286 0.74032511  0.997719
##   ENSBTAG00000011528  0.01183926 0.91556541  0.997719
##   ENSBTAG00000012594 -0.00661189 0.94305715  0.997719
##   -------
##   seqinfo: 48 sequences from UMD3.1 genome
## 
## ...
## &lt;11 more elements&gt;</code></pre>
<p>The resulting <code>DataFrame</code> contains in the first column the CNV regions tested. The second column contains the genes tested in the genomic window around each CNV region, and the other columns report (i) log2 fold change with respect to the 2n group, (ii) edgeR’s DE <em>p</em>-value, and (iii) the (per default) Benjamini-Hochberg adjusted <em>p</em>-value.</p>
</div>
<div id="working-with-tcga-data-stored-in-a-multiassayexperiment" class="section level4">
<h4 class="hasAnchor">
<a href="#working-with-tcga-data-stored-in-a-multiassayexperiment" class="anchor"></a>Working with TCGA data stored in a <code>MultiAssayExperiment</code>
</h4>
<p>In the previous section, we individually prepared the CNV and RNA-seq data for CNV-expression association analysis. In the following, we demonstrate how to perform an integrated preparation of the two assays when stored in a <a href="http://bioconductor.org/packages/MultiAssayExperiment">MultiAssayExperiment</a> We therefore consider glioblastoma <a href="https://cancergenome.nih.gov/cancersselected/glioblastomamultiforme">GBM</a> data from The Cancer Genome Atlas <a href="https://cancergenome.nih.gov">TCGA</a>, which can conveniently be accessed with the <a href="http://bioconductor.org/packages/curatedTCGAData">curatedTCGAData</a> package.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">curatedTCGAData</span><span class="op">)</span>
<span class="va">gbm</span> <span class="op">&lt;-</span> <span class="fu">curatedTCGAData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/curatedTCGAData/man/curatedTCGAData.html">curatedTCGAData</a></span><span class="op">(</span><span class="st">"GBM"</span>,
        assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GISTIC_Peaks"</span>, <span class="st">"CNVSNP"</span>, <span class="st">"RNASeq2GeneNorm"</span><span class="op">)</span>,
        dry.run<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">gbm</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 3 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 3:
##  [1] GBM_CNVSNP-20160128: RaggedExperiment with 146852 rows and 1104 columns
##  [2] GBM_RNASeq2GeneNorm-20160128: SummarizedExperiment with 20501 rows and 166 columns
##  [3] GBM_GISTIC_Peaks-20160128: RangedSummarizedExperiment with 68 rows and 577 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save all data to files</code></pre>
<p>The returned <code>MultiAssayExperiment</code> contains three assays:</p>
<ul>
<li>the SNP-based CNV calls stored in a <code>RaggedExperiment</code> (<code>GBM_CNVSNP</code>),</li>
<li>the recurrent CNV regions summarized across the population using the <a href="http://www.broadinstitute.org/cancer/cga/gistic">GISTIC</a> method (<code>GBM_GISTIC_Peaks</code>), and</li>
<li>the normalized RNA-seq gene expression values in a <code>SummarizedExperiment</code> (<code>GBM_RNASeq2GeneNorm</code>).</li>
</ul>
<p>To annotate the genomic coordinates of the genes measured in the RNA-seq assay, we use the function <code>symbolsToRanges</code> from the <a href="http://bioconductor.org/packages/TCGAutils">TCGAutils</a> package. For demonstration, we restrict the analysis to chromosome 1 and 2.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TCGAutils</span><span class="op">)</span>
<span class="va">gbm</span> <span class="op">&lt;-</span> <span class="fu">TCGAutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TCGAutils/man/simplifyTCGA.html">symbolsToRanges</a></span><span class="op">(</span><span class="va">gbm</span>, unmapped<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>
<span class="op">{</span>
    <span class="va">rr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
    <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">genome</a></span><span class="op">(</span><span class="va">rr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"NCBI37"</span>
    <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html">seqlevelsStyle</a></span><span class="op">(</span><span class="va">rr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"UCSC"</span>
    <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">rr</span><span class="op">)</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rr</span>
    <span class="va">gbm</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gbm</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">ind</span>,<span class="op">]</span>
<span class="op">}</span>
<span class="va">gbm</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 3 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 3:
##  [1] GBM_CNVSNP-20160128: RaggedExperiment with 17818 rows and 1104 columns
##  [2] GBM_GISTIC_Peaks-20160128: RangedSummarizedExperiment with 12 rows and 577 columns
##  [3] GBM_RNASeq2GeneNorm-20160128_ranged: RangedSummarizedExperiment with 2898 rows and 166 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save all data to files</code></pre>
<p>We now restrict the analysis to intersecting patients of the three assays using <code>MultiAssayExperiment</code>’s <code>intersectColumns</code> function, and select <em>Primary Solid Tumor</em> samples using the <code>splitAssays</code> function from the <code>TCGAutils</code> package.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gbm</span> <span class="op">&lt;-</span> <span class="fu">MultiAssayExperiment</span><span class="fu">::</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-helpers.html">intersectColumns</a></span><span class="op">(</span><span class="va">gbm</span><span class="op">)</span>
<span class="fu">TCGAutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TCGAutils/man/curatedTCGAData-helpers.html">sampleTables</a></span><span class="op">(</span><span class="va">gbm</span><span class="op">)</span></code></pre></div>
<pre><code>## $`GBM_CNVSNP-20160128`
## 
##  01  02  10  11 
## 154  13 146   1 
## 
## $`GBM_GISTIC_Peaks-20160128`
## 
##  01 
## 154 
## 
## $`GBM_RNASeq2GeneNorm-20160128_ranged`
## 
##  01  02 
## 147  13</code></pre>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">sampleTypes</span>, package<span class="op">=</span><span class="st">"TCGAutils"</span><span class="op">)</span>
<span class="va">sampleTypes</span></code></pre></div>
<pre><code>##    Code                                        Definition Short.Letter.Code
## 1    01                               Primary Solid Tumor                TP
## 2    02                             Recurrent Solid Tumor                TR
## 3    03   Primary Blood Derived Cancer - Peripheral Blood                TB
## 4    04      Recurrent Blood Derived Cancer - Bone Marrow              TRBM
## 5    05                          Additional - New Primary               TAP
## 6    06                                        Metastatic                TM
## 7    07                             Additional Metastatic               TAM
## 8    08                        Human Tumor Original Cells              THOC
## 9    09        Primary Blood Derived Cancer - Bone Marrow               TBM
## 10   10                              Blood Derived Normal                NB
## 11   11                               Solid Tissue Normal                NT
## 12   12                                Buccal Cell Normal               NBC
## 13   13                           EBV Immortalized Normal              NEBV
## 14   14                                Bone Marrow Normal               NBM
## 15   15                                    sample type 15              15SH
## 16   16                                    sample type 16              16SH
## 17   20                                   Control Analyte             CELLC
## 18   40 Recurrent Blood Derived Cancer - Peripheral Blood               TRB
## 19   50                                        Cell Lines              CELL
## 20   60                          Primary Xenograft Tissue                XP
## 21   61                Cell Line Derived Xenograft Tissue               XCL
## 22   99                                    sample type 99              99SH</code></pre>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gbm</span> <span class="op">&lt;-</span> <span class="fu">TCGAutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TCGAutils/man/curatedTCGAData-helpers.html">splitAssays</a></span><span class="op">(</span><span class="va">gbm</span>, sampleCodes<span class="op">=</span><span class="st">"01"</span><span class="op">)</span>
<span class="va">gbm</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 3 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 3:
##  [1] 01_GBM_CNVSNP-20160128: RaggedExperiment with 17818 rows and 154 columns
##  [2] 01_GBM_GISTIC_Peaks-20160128: RangedSummarizedExperiment with 12 rows and 154 columns
##  [3] 01_GBM_RNASeq2GeneNorm-20160128_ranged: RangedSummarizedExperiment with 2898 rows and 147 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save all data to files</code></pre>
<p>The SNP-based CNV calls from TCGA are provided as segmented log2 copy number ratios.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"CNVSNP"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gbm</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##   Num_Probes Segment_Mean
##    &lt;numeric&gt;    &lt;numeric&gt;
## 1        166       0.1112
## 2          3      -1.2062
## 3      40303       0.1086
## 4        271      -0.3065
## 5      88288       0.1049
## 6      33125       0.3510</code></pre>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">Segment_Mean</span> <span class="op">)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## -8.2199 -0.9779 -0.0035 -0.6395  0.0493  6.9689</code></pre>
<p>It is thus necessary to convert them to integer copy number states for further analysis with <code>CNVRanger</code>.</p>
<p>In a diploid genome, a single-copy gain in a perfectly pure, homogeneous sample has a copy ratio of 3/2. On log2 scale, this is log2(3/2) = 0.585, and a single-copy loss is log2(1/2) = -1.0.</p>
<p>We can roughly convert a log ratio <code>lr</code> to an integer copy number by</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span> <span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="va">lr</span><span class="op">)</span> <span class="op">*</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>Note that this is not the ideal way to calculate absolute integer copy numbers. Especially in cancer, differences in tumor purity, tumor ploidy, and subclonality can substantially interfere with the assumption of a pure homogeneous sample. See <a href="https://doi.org/10.1038/nbt.2203">ABSOLUTE</a> and the <a href="http://bioconductor.org/packages/PureCN">PureCN</a> package for accurately taking such tumor characteristics into account.</p>
<p>However, without additional information we transform the log ratios into integer copy number states using the rough approximation outlined above.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">smean</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">Segment_Mean</span>
<span class="va">state</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="va">smean</span> <span class="op">*</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">state</span><span class="op">[</span><span class="va">state</span> <span class="op">&gt;</span> <span class="fl">4</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">4</span>
<span class="fu">mcols</span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">state</span> <span class="op">&lt;-</span> <span class="va">state</span>
<span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">state</span> <span class="op">!=</span> <span class="fl">2</span>,<span class="op">]</span>
<span class="fu">mcols</span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gbm</span><span class="op">[[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">state</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##    0    1    3    4 
## 2401 4084 1005  747</code></pre>
<p>The data is now ready for CNV-expression association analysis, where we find only four CNV regions with sufficient sample size for testing using the default value of 10 for the <code>minSamples</code> argument.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CNVRanger/man/cnvEQTL.html">cnvEQTL</a></span><span class="op">(</span>cnvrs <span class="op">=</span> <span class="st">"01_GBM_GISTIC_Peaks-20160128"</span>, 
               calls <span class="op">=</span> <span class="st">"01_GBM_CNVSNP-20160128"</span>, 
               rcounts <span class="op">=</span> <span class="st">"01_GBM_RNASeq2GeneNorm-20160128_ranged"</span>, 
               data <span class="op">=</span> <span class="va">gbm</span>, window <span class="op">=</span> <span class="st">"1Mbp"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## harmonizing input:
##   removing 154 sampleMap rows not in names(experiments)</code></pre>
<pre><code>## Preprocessing RNA-seq data ...</code></pre>
<pre><code>## Summarizing per-sample CN state in each CNV region</code></pre>
<pre><code>## Excluding 4 cnvrs not satisfying min.samples threshold</code></pre>
<pre><code>## Analyzing 8 regions with &gt;=1 gene in the given window</code></pre>
<pre><code>## 1 of 8</code></pre>
<pre><code>## 2 of 8</code></pre>
<pre><code>## 3 of 8</code></pre>
<pre><code>## 4 of 8</code></pre>
<pre><code>## 5 of 8</code></pre>
<pre><code>## 6 of 8</code></pre>
<pre><code>## 7 of 8</code></pre>
<pre><code>## 8 of 8</code></pre>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span></code></pre></div>
<pre><code>## GRangesList object of length 8:
## $`chr1:3394251-6475685`
## GRanges object with 29 ranges and 5 metadata columns:
##           seqnames          ranges strand |   logFC.CN1 logFC.CN3 logFC.CN4
##              &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##     RPL22     chr1 6245080-6259679      - |   -0.659072        NA        NA
##      ICMT     chr1 6281253-6296044      - |   -0.564615        NA        NA
##     PHF13     chr1 6673756-6684093      + |   -0.620105        NA        NA
##    KLHL21     chr1 6650779-6662929      - |   -0.695353        NA        NA
##   DNAJC11     chr1 6694228-6761966      - |   -0.553808        NA        NA
##       ...      ...             ...    ... .         ...       ...       ...
##    PRDM16     chr1 2985742-3355185      + |  0.07347117        NA        NA
##     PLCH2     chr1 2407754-2436969      + | -0.09499204        NA        NA
##      HES5     chr1 2460184-2461684      - | -0.07159920        NA        NA
##    KCNAB2     chr1 6052358-6161253      + |  0.02173889        NA        NA
##     AJAP1     chr1 4715105-4843851      + | -0.00693493        NA        NA
##                PValue   AdjPValue
##             &lt;numeric&gt;   &lt;numeric&gt;
##     RPL22 3.20935e-13 3.80537e-12
##      ICMT 5.86039e-11 6.48550e-10
##     PHF13 7.96103e-10 6.95542e-09
##    KLHL21 4.17763e-08 2.77395e-07
##   DNAJC11 9.15183e-08 5.84309e-07
##       ...         ...         ...
##    PRDM16    0.801275    0.841760
##     PLCH2    0.811335    0.841760
##      HES5    0.865276    0.886641
##    KCNAB2    0.922008    0.933252
##     AJAP1    0.982261    0.982261
##   -------
##   seqinfo: 25 sequences (1 circular) from NCBI37 genome
## 
## ...
## &lt;7 more elements&gt;</code></pre>
<p>We can illustrate differential expression of genes in the neighborhood of a CNV region of interest using the function <code>plotEQTL</code>.</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<pre><code>## GRangesList object of length 1:
## $`chr1:7908902-8336254`
## GRanges object with 10 ranges and 5 metadata columns:
##           seqnames          ranges strand | logFC.CN1 logFC.CN3 logFC.CN4
##              &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##     PARK7     chr1 8021714-8045342      + | -0.635321        NA        NA
##      RERE     chr1 8412464-8877699      - | -0.728795        NA        NA
##    CAMTA1     chr1 6845384-7829766      + | -0.575217        NA        NA
##   SLC45A1     chr1 8384390-8404227      + | -0.649802        NA        NA
##      ENO1     chr1 8921059-8939151      - | -0.450498        NA        NA
##     VAMP3     chr1 7831329-7841492      + | -0.394265        NA        NA
##      PER3     chr1 7844380-7905237      + | -0.508559        NA        NA
##      H6PD     chr1 9294863-9331394      + | -0.373110        NA        NA
##    ERRFI1     chr1 8071779-8086393      - | -0.401071        NA        NA
##    SLC2A5     chr1 9097005-9148510      - | -0.275506        NA        NA
##                PValue   AdjPValue
##             &lt;numeric&gt;   &lt;numeric&gt;
##     PARK7 1.85213e-09 1.46406e-08
##      RERE 3.95635e-09 2.98525e-08
##    CAMTA1 1.14737e-08 8.28100e-08
##   SLC45A1 3.06791e-06 1.54325e-05
##      ENO1 4.89692e-04 1.62578e-03
##     VAMP3 7.59027e-04 2.24997e-03
##      PER3 1.71399e-02 3.51263e-02
##      H6PD 2.17771e-02 4.30357e-02
##    ERRFI1 1.10773e-01 1.75127e-01
##    SLC2A5 2.40015e-01 3.23922e-01
##   -------
##   seqinfo: 25 sequences (1 circular) from NCBI37 genome</code></pre>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 1 range and 0 metadata columns:
##       seqnames          ranges strand
##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1 7908902-8336254      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/CNVRanger/man/plotEQTL.html">plotEQTL</a></span><span class="op">(</span>cnvr<span class="op">=</span><span class="va">r</span>, genes<span class="op">=</span><span class="va">res</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, genome<span class="op">=</span><span class="st">"hg19"</span>, cn<span class="op">=</span><span class="st">"CN1"</span><span class="op">)</span></code></pre></div>
<p><img src="Geistlinger_CNVWorkshop_files/figure-html/plotEQTL-1.png" width="100%"></p>
<p>The plot shows consistent decreased expression (negative log2 fold change) of genes in the neighborhood of the CNV region, when comparing samples with a one copy loss (1<span class="math inline">\(n\)</span>) in that region to the 2<span class="math inline">\(n\)</span> reference group.</p>
<p>Note that a significant expression change is not only observed for genes locating within the CNV region (dosage effect, here: PARK7), but also genes locating in close proximity of the CNV region (neighborhood effect, here: CAMTA1 and RERE). This is consistent with previous observations in mouse <a href="https://doi.org/10.1038/ng.350">Cahan et al., 2009</a> and our observations in cattle <a href="https://doi.org/10.1038/s41598-018-19782-4">Geistlinger et al., 2018</a>.</p>
</div>
</div>
</div>
<div id="allele-specific-cnv-analysis-with-purecn" class="section level2">
<h2 class="hasAnchor">
<a href="#allele-specific-cnv-analysis-with-purecn" class="anchor"></a>Allele-specific CNV analysis with PureCN</h2>
<p>For allele-specific copy number alteration (CNA) analysis, most commonly used tools in the field rely on high quality genome-wide data with matched normal profiles, limiting their applicability in clinical settings.</p>
<p>Here, we’re introducing the open-source <a href="http://bioconductor.org/packages/PureCN">PureCN</a> R/Bioconductor package in conjunction with widely used variant-calling and copy number segmentation algorithms, for allele-specific CNA analysis from whole exome sequencing (WES) <em><strong>without matched normals</strong></em>.</p>
<p>Further information on PureCN and its CNV analysis workflow can be found here:<br>
- <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/PureCN/inst/doc/PureCN.pdf">PureCN vignette</a><br>
- <a href="https://doi.org/10.1186/s13029-016-0060-z">PureCN paper</a><br>
- <a href="https://ascopubs.org/doi/10.1200/CCI.19.00130">Reliable analysis of clinical tumor-only whole exome sequencing data</a> (Oh <em>et al.</em>, JCO Clin Cancer Inform. 2020 Apr;4:321-335. doi: 10.1200/CCI.19.00130.)<br>
- <a href="https://app.terra.bio/#workspaces/waldronlab-terra/Tumor_Only_CNV">Terra Workspace: Tumor_Only_CNV</a> (<a href="http://waldronlab.io/AnVILWorkshop/">Related workshop</a> on cloud-based genomics platform, Terra is available during 9-10am on Thursday July 30th)<br>
- Detailed scripts are available in this <a href="https://github.com/shbrief/CNVWorkflow_Code">GitHub repo</a>.</p>
<div id="prepare-input-files-using-purecn-and-mutect" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-input-files-using-purecn-and-mutect" class="anchor"></a>Prepare input files using PureCN and MuTect</h3>
<p>PureCN requires three raw inputs and subsequent preprocessing of them before performing purity and ploidy estimation. Here are the list of three raw input files, followed by the simple workflow diagram.</p>
<div id="bed-file" class="section level5">
<h5 class="hasAnchor">
<a href="#bed-file" class="anchor"></a>1. BED file</h5>
<p>we are analyzing WES data, so we need to know which part of the genome is captured through BED file.</p>
</div>
<div id="process-matched-normal-bam-files" class="section level5">
<h5 class="hasAnchor">
<a href="#process-matched-normal-bam-files" class="anchor"></a>2. <em><strong>Process-matched</strong></em> normal BAM files</h5>
<p>these allow us to find out what kinds of backgrounds are there, especially technical biases.</p>
</div>
<div id="tumor-bam-files" class="section level5">
<h5 class="hasAnchor">
<a href="#tumor-bam-files" class="anchor"></a>3. Tumor BAM files</h5>
<p>these contain the genetic changes, both germline and somatic alternations.<br><br></p>
<p><img src="overview_1.png" width="40%"></p>
<p><br></p>
</div>
</div>
<div id="five-intermediate-files" class="section level3">
<h3 class="hasAnchor">
<a href="#five-intermediate-files" class="anchor"></a>Five intermediate files</h3>
<p>Preprocessing of three input files creates five major intermediate files that are used directly for purity and ploidy estimation with <code>PureCN.R</code> command line tool.</p>
<div id="interval-file" class="section level5">
<h5 class="hasAnchor">
<a href="#interval-file" class="anchor"></a>1. Interval file</h5>
<p>IntervalFile.R command line tool optimize the targets for copy number calling, for example,</p>
<ul>
<li>large targets are split to obtain higher resolution<br>
</li>
<li>targets in regions of low mappability are dropped</li>
</ul>
<p>One thing you should be careful is to make sure that the genome version of target file matches the reference.</p>
</div>
<div id="vcf-and-stats-files" class="section level5">
<h5 class="hasAnchor">
<a href="#vcf-and-stats-files" class="anchor"></a>2. VCF and stats files</h5>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1"></a>java <span class="op">-</span>jar mutect.jar \</span>
<span id="cb145-2"><a href="#cb145-2"></a>    <span class="op">--</span>analysis_type MuTect \</span>
<span id="cb145-3"><a href="#cb145-3"></a>    <span class="op">-</span>R hg38.fasta \</span>
<span id="cb145-4"><a href="#cb145-4"></a>    <span class="op">--</span>dbsnp <span class="op">$</span>DBSNP_VCF \</span>
<span id="cb145-5"><a href="#cb145-5"></a>    <span class="op">--</span>cosmic <span class="op">$</span>COSMIC_VCF \</span>
<span id="cb145-6"><a href="#cb145-6"></a>    <span class="op">-</span>I<span class="op">:</span>tumor <span class="op">$</span>BAM_TUMOR  \</span>
<span id="cb145-7"><a href="#cb145-7"></a>    <span class="op">-</span>o <span class="op">$</span>OUT<span class="op">/</span><span class="er">$</span>{SAMPLEID}_mutect_stats.txt \</span>
<span id="cb145-8"><a href="#cb145-8"></a>    <span class="op">-</span>vcf <span class="op">$</span>OUT<span class="op">/</span><span class="er">$</span>{SAMPLEID}_mutect.vcf</span></code></pre></div>
<p>Output VCF files contain both germline SNPs and somatic mutations, which have information about read-depths of reference and alt alleles and whether the variant is in dbSNP. <code>call_stats.txt</code> file is an through report of all the metrics and statistics available about the calls made my MuTect and the filtered that are applied internally by default. With this stats file, PureCN will keep germline variants, while removing potential artifacts.</p>
</div>
<div id="pool-of-normals-pon" class="section level5">
<h5 class="hasAnchor">
<a href="#pool-of-normals-pon" class="anchor"></a>3. Pool of normals (PoN)</h5>
<p>For this workflow, MuTect is run on <code>--artifact-detection-mode</code>,</p>
<ul>
<li>which is used when running the caller on a normal (as if it were a tumor) to detect artifacts</li>
<li>which includes variant calls that are clearly germline</li>
<li>which is commonly used for Panel-Of-Normals creation</li>
</ul>
<div class="sourceCode" id="cb146"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb146-1"><a href="#cb146-1"></a><span class="ex">java</span> -jar mutect.jar \</span>
<span id="cb146-2"><a href="#cb146-2"></a>   --analysis_type MuTect \</span>
<span id="cb146-3"><a href="#cb146-3"></a>   -R hg38.fasta \</span>
<span id="cb146-4"><a href="#cb146-4"></a>   --artifact_detection_mode \</span>
<span id="cb146-5"><a href="#cb146-5"></a>   --dbsnp <span class="va">$DBSNP_VCF</span> \</span>
<span id="cb146-6"><a href="#cb146-6"></a>   --cosmic <span class="va">$COSMIC_VCF</span> \</span>
<span id="cb146-7"><a href="#cb146-7"></a>   -dt None \</span>
<span id="cb146-8"><a href="#cb146-8"></a>   -I:tumor <span class="va">$BAM_NORMAL</span>  \</span>
<span id="cb146-9"><a href="#cb146-9"></a>   -o <span class="va">$OUT</span>/<span class="va">${SAMPLEID}</span>_pon_stats.txt \</span>
<span id="cb146-10"><a href="#cb146-10"></a>   -vcf <span class="va">$OUT</span>/<span class="va">${SAMPLEID}</span>_pon.vcf</span></code></pre></div>
<p>Depending on the type of variant you’re looking for, the PON will be generated differently. What all PONs have in common is that,</p>
<ul>
<li>they are made from normal samples (in this context, “normal” means derived from healthy tissue that is believed to not have any somatic alterations) and</li>
<li>their main purpose is to capture recurrent technical artifacts in order to improve the results of the variant calling analysis.</li>
</ul>
</div>
<div id="gc-normalized-coverage-files" class="section level5">
<h5 class="hasAnchor">
<a href="#gc-normalized-coverage-files" class="anchor"></a>4. GC-normalized coverage files</h5>
<p><code>Coverage.R</code> process tumor and normal bam files to calculate GC-normalized coverages. Different library can give different coverage profiles, and the most important library-specific bias is due to GC-content. i.e. regions of high AT- or GC-content are not always captured with exactly the same efficiency in tumor and normals.</p>
</div>
<div id="normaldb" class="section level5">
<h5 class="hasAnchor">
<a href="#normaldb" class="anchor"></a>5. NormalDB</h5>
<p>To build a normalDB for coverage normalization and bias correction, we are providing two intermediate files: 1) the list of coverage normalized files from <code>Coverage.R</code> and 2) normal panel vcf file (PON).</p>
<p>Three outputs from this process are used for PureCN,</p>
<ul>
<li>a database of normal samples, which will be used to normalize tumor coverages,<br>
</li>
<li>mapping bias information of each variant. By examining the coverage of particular intervals in a pool of normals, we can estimate how well this assay captures these intervals and will then adjust the tumor coverage accordingly.<br>
</li>
<li>Last is an interval weight file. Interval weights will be set proportional to the inverse of coverage standard deviation across all normals. Intervals with high variance in coverage in the pool of normals are thus down-weighted.</li>
</ul>
<p><br></p>
</div>
</div>
<div id="run-purecn-r" class="section level3">
<h3 class="hasAnchor">
<a href="#run-purecn-r" class="anchor"></a>Run PureCN.R</h3>
<p>All the pre-processed files above are used as inputs for <code>PureCN.R</code> to normalize, segment and determine purity and ploidy. This step returns purity and ploidy combinations, sorted by likelihood score. Provides copy number and LOH data, by both gene and genomic region.</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1"></a>Rscript <span class="op">$</span>PURECN<span class="op">/</span>PureCN.R \</span>
<span id="cb147-2"><a href="#cb147-2"></a>    <span class="op">--</span>out <span class="op">$</span>PCN_OUT<span class="op">/</span>PureCN<span class="op">/</span>tumor_only<span class="op">/</span><span class="er">$</span>SAMPLEID \</span>
<span id="cb147-3"><a href="#cb147-3"></a>    <span class="op">--</span>tumor <span class="op">$</span>PCN_OUT<span class="op">/</span>tumor_cov<span class="op">/</span><span class="er">$</span>{SAMPLEID}_coverage_loess.txt \</span>
<span id="cb147-4"><a href="#cb147-4"></a>    <span class="op">--</span>SAMPLEID <span class="op">$</span>{SAMPLEID} \</span>
<span id="cb147-5"><a href="#cb147-5"></a>    <span class="op">--</span>vcf <span class="op">$</span>MUTECT_OUT<span class="op">/</span>stat_tumor_only<span class="op">/</span><span class="er">$</span>{SAMPLEID}_mutect.vcf \</span>
<span id="cb147-6"><a href="#cb147-6"></a>    <span class="op">--</span>statsfile <span class="op">$</span>MUTECT_OUT<span class="op">/</span>stat_tumor_only<span class="op">/</span><span class="er">$</span>{SAMPLEID}_mutect_stats.txt \</span>
<span id="cb147-7"><a href="#cb147-7"></a>    <span class="op">--</span>normaldb <span class="op">$</span>PCN_OUT<span class="op">/</span>normalDB<span class="op">/</span>normalDB_hg38.rds \</span>
<span id="cb147-8"><a href="#cb147-8"></a>    <span class="op">--</span>normal_panel <span class="op">$</span>PCN_OUT<span class="op">/</span>normalDB<span class="op">/</span>mapping_bias_hg38.rds \</span>
<span id="cb147-9"><a href="#cb147-9"></a>    <span class="op">--</span>intervals <span class="op">$</span>INPUT<span class="op">/</span>bed<span class="op">/</span>baits_hg38_intervals.txt \</span>
<span id="cb147-10"><a href="#cb147-10"></a>    <span class="op">--</span>intervalweightfile <span class="op">$</span>PCN_OUT<span class="op">/</span>normalDB<span class="op">/</span>interval_weights_hg38.txt \</span>
<span id="cb147-11"><a href="#cb147-11"></a>    <span class="op">--</span>snpblacklist hg38_simpleRepeats.bed \</span>
<span id="cb147-12"><a href="#cb147-12"></a>    <span class="op">--</span>genome hg38 \</span>
<span id="cb147-13"><a href="#cb147-13"></a>    <span class="op">--</span>force <span class="op">--</span>postoptimize <span class="op">--</span>seed <span class="dv">123</span></span></code></pre></div>
<p><br></p>
<p><strong>Overview of CNV analysis workflow with intermediates</strong><br>
Here is the more detailed workflow diagram with the intermediate files described above. <img src="overview_2.png" width="130%"></p>
<p><br></p>
<div id="for-production" class="section level4">
<h4 class="hasAnchor">
<a href="#for-production" class="anchor"></a>For production</h4>
<p>If you are collecting tumor samples using the same protocol (with the same capture kit and sequencing read length), you need to process BED and process-matched normal samples only once to build a normal database. Here, the ‘normal database’ includes <code>normalDB</code>, <code>normal_panel</code>, <code>intervals</code>, <code>intervalweightfile</code>, <code>snpblacklist</code> files. At the end, you need to run three workflows for each new tumor samples: <em>1_MuTect1_Variants_Calling</em>, <em>2_PureCN_Coverage</em>, and <em>4_PureCN_PureCN</em> workflows.</p>
<p><strong>Requirement for production pipeline</strong><br>
As long as you are using the same capture kit and the same sequencing protocols, you need to process only tumor bam files for the production. <img src="overview_3.png" width="130%"></p>
</div>
</div>
<div id="analysis-examples" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-examples" class="anchor"></a>Analysis examples</h3>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lima1/PureCN">PureCN</a></span><span class="op">)</span>
<span class="va">file.rds</span> <span class="op">=</span> <span class="st">"Sample1_PureCN.rds"</span>
<span class="va">ret</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">file.rds</span><span class="op">)</span></code></pre></div>
<div id="overview-1" class="section level5">
<h5 class="hasAnchor">
<a href="#overview-1" class="anchor"></a>1. Overview</h5>
<p>The colors visualize the copy number fitting score from low (blue) to high (red). The numbers indicate the ranks of the local optima.</p>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/PureCN/man/plotAbs.html">plotAbs</a></span><span class="op">(</span><span class="va">ret</span>, type<span class="op">=</span><span class="st">"overview"</span><span class="op">)</span></code></pre></div>
<p><img src="Geistlinger_CNVWorkshop_files/figure-html/unnamed-chunk-27-1.png" width="60%"></p>
</div>
<div id="log-ratio-histogram" class="section level5">
<h5 class="hasAnchor">
<a href="#log-ratio-histogram" class="anchor"></a>2. Log-ratio histogram</h5>
<p>This figure displays a histogram of tumor vs. normal copy number log-ratios for the maximum likelihood solution (number 1 of ‘Overview’ plot). The height of a bar in this plot is proportional to the fraction of the genome falling into the particular log-ratio copy number range.</p>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/PureCN/man/plotAbs.html">plotAbs</a></span><span class="op">(</span><span class="va">ret</span>, <span class="fl">1</span>, type<span class="op">=</span><span class="st">"hist"</span><span class="op">)</span></code></pre></div>
<p><img src="Geistlinger_CNVWorkshop_files/figure-html/unnamed-chunk-28-1.png" width="60%"></p>
<pre><code>## NULL</code></pre>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/PureCN/man/createCurationFile.html">createCurationFile</a></span><span class="op">(</span><span class="va">file.rds</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"Sample1_PureCN.csv"</span><span class="op">)</span></code></pre></div>
<pre><code>##   Sampleid Purity   Ploidy Sex Contamination Flagged Failed Curated
## 1  Sample1   0.65 1.734572   ?             0    TRUE  FALSE   FALSE
##                          Comment
## 1 EXCESSIVE LOSSES;EXCESSIVE LOH</code></pre>
</div>
<div id="amplification-and-deletion" class="section level5">
<h5 class="hasAnchor">
<a href="#amplification-and-deletion" class="anchor"></a>3. Amplification and deletion</h5>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene.calls</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/PureCN/man/callAlterations.html">callAlterations</a></span><span class="op">(</span><span class="va">ret</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gene.calls</span><span class="op">)</span></code></pre></div>
<pre><code>##            chr     start       end C seg.mean seg.id number.targets  gene.mean
## EIF2A     chr3 150270143 150301699 6   1.3351      5             13  1.5159387
## MIR548H2  chr3 151531951 151538241 6   1.3351      5              3  0.8261594
## GPNMB     chr7  23286477  23313844 7   1.4723     17             11  1.4490939
## SH2D4B   chr10  82298088  82403838 0  -1.5118     22              8 -1.5141982
## SLC35G1  chr10  95653791  95661248 0  -1.3002     24              3 -1.5988526
## CYP2C9   chr10  96698440  96748786 0  -1.3002     24              9 -1.1901085
##            gene.min   gene.max focal breakpoints          type num.snps  M
## EIF2A     0.7544062  2.1392167  TRUE           0 AMPLIFICATION        0 NA
## MIR548H2  0.4682628  1.2000253  TRUE           0 AMPLIFICATION        0 NA
## GPNMB     0.9397072  1.8200775  TRUE           0 AMPLIFICATION        0 NA
## SH2D4B   -1.9281007 -1.2607230 FALSE           0      DELETION        2  0
## SLC35G1  -1.8066816 -1.4493561 FALSE           0      DELETION        0 NA
## CYP2C9   -2.1922992 -0.9875222 FALSE           0      DELETION        0 NA
##          M.flagged  loh
## EIF2A           NA   NA
## MIR548H2        NA   NA
## GPNMB           NA   NA
## SH2D4B        TRUE TRUE
## SLC35G1         NA   NA
## CYP2C9          NA   NA</code></pre>
</div>
<div id="genomic-regions-with-loh" class="section level5">
<h5 class="hasAnchor">
<a href="#genomic-regions-with-loh" class="anchor"></a>4. Genomic regions with LOH</h5>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">loh</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/PureCN/man/callLOH.html">callLOH</a></span><span class="op">(</span><span class="va">ret</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">loh</span><span class="op">)</span></code></pre></div>
<pre><code>##    chr     start       end arm C M                       type   seg.mean
## 1 chr1 114515871 121535434   p 2 0 WHOLE ARM COPY-NEUTRAL LOH  0.1209814
## 2 chr1 124535434 248085104   q 2 0 WHOLE ARM COPY-NEUTRAL LOH  0.1209814
## 3 chr2  10262881  92326171   p 1 0              WHOLE ARM LOH -0.4656416
## 4 chr2  95326171 231775678   q 1 0              WHOLE ARM LOH -0.4656416
## 5 chr2 236403331 239039169   q 2 0           COPY-NEUTRAL LOH  0.0193000
## 6 chr3  11888043  90504854   p 2 1                             0.1052000
##   num.mark num.snps M.flagged maf.expected maf.observed
## 1      902       13     FALSE    0.1750000    0.1756325
## 2      902       13     FALSE    0.1750000    0.1756325
## 3      685       13     FALSE    0.2592593    0.2788990
## 4      685       13     FALSE    0.2592593    0.2788990
## 5       88        2      TRUE    0.1750000    0.1736667
## 6      408        1      TRUE    0.5000000    0.4712710</code></pre>
</div>
</div>
<div id="example-of-tcga-ov-and-tcga-luad-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#example-of-tcga-ov-and-tcga-luad-datasets" class="anchor"></a>Example of TCGA-OV and TCGA-LUAD datasets</h3>
<p><strong>1. Purity and Ploidy</strong><br><strong>2. TMB (Tumor mutation burden)</strong> : TMB is the number of somatic mutations carried by tumor cells.<br><strong>3. Mutational signatures</strong> : Mutational signatures are characteristic combinations of mutation types arising from specific mutagenesis processes, which can provide information on cause and treatment option for different types of cancer.</p>
<p><img src="analysis_example.png" width="100%"></p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>A <em>disjoint</em> set of ranges has no overlap between any ranges of the set.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
